drivers/spi/spi_ll_stm32.c

```diff
@@ -887,34 +887,73 @@ static inline bool spi_stm32_is_subghzspi(const struct device *dev)
 #endif
 }
 
-static int spi_stm32_init(const struct device *dev)
+static inline int spi_stm32_clock_configure(const struct spi_stm32_config *cfg)
 {
-	struct spi_stm32_data *data __attribute__((unused)) = dev->data;
-	const struct spi_stm32_config *cfg = dev->config;
 	int err;
 
-	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
-		LOG_ERR("clock control device not ready");
-		return -ENODEV;
-	}
-
 	err = clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
-			       (clock_control_subsys_t) &cfg->pclken[0]);
+			       (clock_control_subsys_t)&cfg->pclken[0]);
 	if (err < 0) {
 		LOG_ERR("Could not enable SPI clock");
 		return err;
 	}
 
 	if (IS_ENABLED(STM32_SPI_DOMAIN_CLOCK_SUPPORT) && (cfg->pclk_len > 1)) {
 		err = clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
-					      (clock_control_subsys_t) &cfg->pclken[1],
-					      NULL);
+					      (clock_control_subsys_t)&cfg->pclken[1], NULL);
 		if (err < 0) {
 			LOG_ERR("Could not select SPI domain clock");
 			return err;
 		}
 	}
 
+	return 0;
+}
+
+/**
+ * @brief Reinitialization of SPI context
+ *
+ * This function reenables clocks, which is required upon exiting certain
+ * low-power modes on select SoCs.
+ *
+ * @param dev SPI device struct
+ *
+ * @return 0
+ */
+int spi_stm32_reinit(const struct device *dev)
+{
+	const struct spi_stm32_config *cfg = dev->config;
+	int err;
+
+	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	err = spi_stm32_clock_configure(cfg);
+	if (err < 0) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int spi_stm32_init(const struct device *dev)
+{
+	struct spi_stm32_data *data __attribute__((unused)) = dev->data;
+	const struct spi_stm32_config *cfg = dev->config;
+	int err;
+
+	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	err = spi_stm32_clock_configure(cfg);
+	if (err < 0) {
+		return err;
+	}
```

Comment:
this is introducing code duplication. what is really required after stop2? only clock re-enablement or state recovery as well?

Reply:
Done.